### PR TITLE
add structure.id

### LINKED ISSUE #3: Repurpose the structure category
Please see discussion at https://github.com/COMCIFS/cif_core/pull/445#issue-1796323136 for background.

Original approach:

1. All categories describing the space group are assigned a `_space_group.id` key data name or pointer. This makes explicit the fact that they all relate to the same space group.
    
 2. The STRUCTURE category now holds a description of structure, which consists of a pointer to the space group and to the experimental conditions (`_diffrn.id`) that it relates to.
    
3. The ATOM_SITE and CELL categories now include a key data name pointer to `_structure.id`, identifying which structure and experimental conditions an atom or cell relates to.
    
4. More controversially, `_cell.diffrn_id` has been replaced by `_structure.diffrn_id`. To strictly follow our rules, it should be deprecated, but it is unlikely to have been used since the release.

If @vaitkus and @rowlesmr could confirm that the above discussion satisfied their objections, I will recreate the pull request.

### PR DESCRIPTION
**Pull request by rowlesmr**
Will close #3 

I redid the PR, as I wanted to work through things in my head.

There are now:

- `STRUCTURE`
   - `id`
   - `space_group_id`
   - `diffrn_id`
- `ATOM_SITE`, `ATOM_SITE_ANISO`, `CELL`, `CELL_MEASUREMENT`, and `CELL_MEASUREMENT_REFLN`
   - `structure_id`
- `SPACE_GROUP`
   - `id`
- `SPACE_GROUP_GENERATOR`, `SPACE_GROUP_SYMOP`, and `SPACE_GROUP_WYCKOFF`
   - `space_group_id`

`cell.diffrn_id` and `cell_measurement.diffrn_id` have been removed.

Modulated and magnetic structures need to be looked at.

### CODE REVIEW

**Comment by jamesrhester** [comment]
That example does demonstrate exactly how I imagine this working.

File: multi_block_core.dic [hunk 1]
```diff
@@ -241,6 +276,388 @@ save_exptl_crystal.id
...
+save_STRUCTURE
+
+    _definition.id                STRUCTURE
+    _definition.scope             Category
+    _definition.class             Set
+    _definition.update            2023-10-16
+    _description.text
+;
+    The DICTIONARY group encompassing the CORE STRUCTURE data items defined
+    and used within the Crystallographic Information Framework (CIF).
```
Comment (jamesrhester) [inline multi_block_core.dic:463]: This category definition needs to be replaced. It is no longer a 'dictionary group' but just an ordinary category. Something like "A category for collecting information about a crystallographic structure"

**Review by jamesrhester — CHANGES_REQUESTED** [review]
As mentioned in a comment to #3, `_cell.diffrn_id` and `_cell_measurement.diffrn_id` should stay but are no longer key data names.

**Comment by vaitkus** [comment]
I suggest that we include the example given in (https://github.com/COMCIFS/MultiBlock_Dictionary/pull/6#issuecomment-1764741476) as distinct CIF files in the PR. This will definitely be useful, since people are already asking for usage examples.

Furthermore, I have a comment on the following statement given in the example :
```
_space_group.id 1 #Must be unique. Can be the same if representing the same space group in the same setting
```
I think that having the same setting if not sufficient. For space groups to have the same identifier, their symmetry operations in the  `SPACE_GROUP` loop must be listed with the same symop ids since these ids are later on used to specify symmetry operations in data items like `_geom_bond.site_symmetry_1`.

Consider the following example:
```
data_merged
# ...
loop_
_space_group.id
_space_group.name_H-M_alt
1 'P 1 21/m 1'
2 'P 1 21/m 1'
# ...
loop_
_space_group_symop.space_group_id
_space_group_symop.id
_space_group_symop.operation_xyz
1 1 x,y,z
1 2 -x,y+1/2,-z
1 3 -x,-y,-z
1 4 x,-y+1/2,z
2 1 x,y,z
2 2 -x,y+1/2,-z
2 3 x,-y+1/2,z
2 4 -x,-y,-z

loop_
_geom_bond.atom_site_label_1
_geom_bond.atom_site_label_2
_geom_bond.space_group_id # not currently defined
_geom_bond.site_symmetry_1
_geom_bond.site_symmetry_2
_geom_bond.distance
C2 C3  1 1_555 3_555 1.44
```
Semantically, the two space groups are identical (same name, same number, same setting, same symmetry operations), but due to the different ids assigned to the symops, they have to retain distinct ids.

Furthermore, `_geom_bond.space_group_id` in the `GEOM_BOND` loop should probably be replaced by `_geom_bond.structure_id`, but this assumes, that the proper structure-to-space-group relationship is defined in the `STRUCTURE` loop.

I do not think that we can achieve a more elegant solution in the constraints of the relational model since items like `_geom_bond.structure_id` prevent normalisation, but we need to be sure to properly communicate such gotchas to the users. Maybe it would make sense to describe the criteria required for two space groups to share the same space group id the definition of the `_space_group.id` data item?

**Comment by rowlesmr** [comment]
That sounds doable. A space group is the same iff it has the same number, setting, and symops in the same order

File: multi_block_core.dic (hunk 1 above)
Comment (rowlesmr) [reply multi_block_core.dic:463]: Done

**Comment by jamesrhester** [comment]
> > As mentioned in a comment to #3, `_cell.diffrn_id` and `_cell_measurement.diffrn_id` should stay but are no longer key data names.
> 
> They already exist in core. The multiblock just alters the key dataname. I take this to mean that they remain in the dictionary, so no need to redefine them?

I think that they should be moved to the multiblock dictionary, after which they will be removed from the core dictionary. This is because these data names have no use in the single-data-block paradigm (you can't refer to a `_diffrn.id` that is not the same as the current data block).

**Comment by jamesrhester** [comment]
> That sounds doable. A space group is the same iff it has the same name, number, setting, and symops in the same order. I can add this to the category description.

Note that this is (mostly) just a particular case of the general rule that "if you repeat key data name values in different blocks, the rest of the values in the row must be identical". I don't think it warrants special mention in the category definition, but is worth pointing out to programmers as it is a useful consistency check. `space_group_symop` is a little special, because we have to autogenerate the symop numbers for legacy files. Anybody writing software now (and thus reading the dictionary) will provide symop ids, so I doubt mentioning this in the core dictionary will do anything except confuse readers.

And I'd say that a space group is the same if the items in the `space_group` category are the same for the same values of the key data name. That is, changing the order of the symops does not change the space group (just like in real life). What is does is change the identity of a symop that belongs to the space group.

**Comment by rowlesmr** [comment]
> Note that this is (mostly) just a particular case of the general rule that "if you repeat key data name values in different blocks, the rest of the values in the row must be identical". I don't think it warrants special mention in the category definition, but is worth pointing out to programmers as it is a useful consistency check. `space_group_symop` is a little special, because we have to autogenerate the symop numbers for legacy files. Anybody writing software now (and thus reading the dictionary) will provide symop ids, so I doubt mentioning this in the core dictionary will do anything except confuse readers.

I haven't put it in the category description; it's in the `_space_group.id` description.

 
> And I'd say that a space group is the same if the items in the `space_group` category are the same for the same values of the key data name. That is, changing the order of the symops does not change the space group (just like in real life). What is does is change the identity of a symop that belongs to the space group.

I agree that changing the order of the symops doesn't change the symmetry, but it does change how the symmetry is represented. As @vaitkus pointed out, having different `_geom_*.symmetry_*` values pointing to different symop id values requires that the symop indicated by that id be the same, and so it doesn't matter the order, as long as the rows have the correct id in the loop, and hence I think it is worth pointing out. It is a gotcha.

File: multi_block_core.dic [hunk 2]
```diff
@@ -147,6 +147,24 @@ save_cell.diffrn_id
...
 save_
 
+save_cell.structure_id
+
+    _definition.id                '_cell.structure_id'
+    _definition.update            2023-10-16
+    _description.text
+;
+    A code identifying the structure to which this atom site belongs.
+;
```
Comment (jamesrhester) [inline multi_block_core.dic:157]: Not 'atom site' but 'cell'. "A code identifying the crystallographic structure associated with this cell description."

File: multi_block_core.dic [hunk 3]
```diff
@@ -208,6 +226,65 @@ save_cell_measurement.diffrn_id
...
 save_
 
+save_cell_measurement.structure_id
+
+    _definition.id                '_cell_measurement.structure_id'
+    _definition.update            2023-10-16
+    _description.text
+;
+    A code identifying the structure to which the cell belongs.
+;
```
Comment (jamesrhester) [inline multi_block_core.dic:236]: "A code identifying the structure to which the cell being measured belongs."

File: multi_block_core.dic [hunk 4]
```diff
@@ -120,14 +120,14 @@ save_CELL
...
 save_
 
 save_cell.diffrn_id
 
     _definition.id                '_cell.diffrn_id'
-    _definition.update            2023-02-01
+    _definition.update            2023-10-26
     _description.text
 ;
     A pointer to the diffraction conditions to which this cell has been applied,
```
Comment (jamesrhester) [inline multi_block_core.dic:133]: `_cell.diffrn_id` is unnecessary as it has the same value as `_structure.diffrn_id`.

**Comment by jamesrhester** [comment]
An analysis of implications of the new `STRUCTURE` category with other dictionaries. I want to understand what we are creating for them by adding linked data names to categories that they also modify.

1. Modulated structures. A modulated or composite structure is described using modulation waves and/or a series of "subsystems" that interpenetrate.
   a.  `Atom_Site`: A subsystem id and flags for the type of modulation wave a particular site is additionally modified by are added, but the modulation information is in a different category
   b. `cell`: modulation information is added in this and other categories
   c. `space group`: superspace group information also added, same for `space_group_symop` etc.

A number of additional categories are defined that provide more modulation information. None of these have been made formal children of the above categories (yet) and so can be ignored - it is up to the ms_dic people to decide if they want to do more.

The `cell_subsystem` category is interesting, as it adopts the same approach as we have with multi blocks. However, only the `atom_site` category explicitly contains a pointer to `_cell_subsystem.code`.

**Conclusion**: 'Structure' as we have defined it (`atom_site` + `cell` + `space_group`) will partially describe a superspace structure. `ms_dic` can completely describe a structure (in the same sense as core CIF is complete) by adding the appropriate links to `_structure.id` in the appropriate categories. A single structure will include all subsystems.

2. Magnetic structures. A magnetic structure is described by the positions and atomic moments of a subset of atoms combined with a magnetic space group. The modulations and superspace groups of the modulated structures dictionary are also used. The magnetic `atom_site_moment` category is a formal child of `atom_site`, meaning that atomic moments are currently incorporated into our concept of `structure`.

Here we come to an important point: we can imagine a `magnetic_structure` category that fulfills the same role as the `structure` category, but for magnetism. Such a category would have a pointer to `_structure.id`, and then include the magnetic space group and magnetism-specific modulation waves. What concerns me (slightly) is that the moments really belong only to the magnetic structure but are swept up into the structure overall. That is a consequence of wanting to list the moments in a single loop with the atomic positions, so in that sense the decision has already been made.  An alternative to a separate magnetic structure category is simply for the magnetic structure dictionary to expand the concept of structure and add magnetic information to the `structure` category.

The magnetism dictionary also has the idea of a parent space group, which is a non-magnetic structure that the magnetic structure is related to. This is a good use for a pointer to `_structure.id`

**Conclusion** :  the structure category is a slight misnomer in the case of magnetism, but does not create practical problems. If we want to be more universal in our naming, we could change the name to something like `STRUCTURAL_MODEL`

**Comment by jamesrhester** [comment]
I think this PR is ready for incorporation into the multi block dictionary following the above suggested changes and perhaps a change of the category from `STRUCTURE` to something like `STRUCTURAL_MODEL` to try to convey the more general meaning. Note that version 1.0.0 of the multi-block dictionary is now available from the IUCr website.

**Review by jamesrhester — APPROVED** [review]

File: multi_block_core.dic [hunk 5]
```diff
@@ -241,6 +296,391 @@ save_exptl_crystal.id
...
+    _type.purpose                 Link
+    _type.source                  Assigned
+    _type.container               Single
+    _type.contents                Word
+
+save_
+
+save_STRUCTURE
+
+    _definition.id                STRUCTURE
```
[the remark this answers is not in the document]
Comment (jamesrhester) [reply multi_block_core.dic:480]: I'd like some feedback on whether or not a new name seems better, given how the category is used.

File: multi_block_core.dic [hunk 6]
```diff
@@ -241,6 +296,391 @@ save_exptl_crystal.id
...
+    loop_
+      _category_key.name
+         '_atom_site.label'
+         '_atom_site.structure_id'
+
+    loop_
+      _description_example.case
+      _description_example.detail
+;
+         loop_
```
[the remark this answers is not in the document]
Comment (jamesrhester) [reply multi_block_core.dic:571]: Let's create a separate issue for this.

File: multi_block_core.dic (hunk 5 above)
[the remark this answers is not in the document]
Comment (jamesrhester) [reply multi_block_core.dic:480]: Having had no feedback I think I'll stick with the more general STRUCTURE. We can change it before the next release if there are objections.

**Comment by jamesrhester** [comment]
Finally merged. The examples may need improvement.

**Pull request merged**